arcgis-ios-sdk-samples/Maps/Generate offline map (basemap by reference)/GenerateOfflineMapBasemapByReferenceViewController.swift

```diff
@@ -0,0 +1,225 @@
+            presentAlert(error: error)
+        } else {
+            generateButtonItem.isEnabled = true
+        }
+    }
+    
+    /// Called in response to the Generate Offline Map button item being tapped.
+    @IBAction func generateOfflineMap() {
+        generateButtonItem.isEnabled = false
+        mapView.isUserInteractionEnabled = false
+        let offlineMapTask = AGSOfflineMapTask(onlineMap: mapView.map!)
+        offlineMapTask.defaultGenerateOfflineMapParameters(withAreaOfInterest: areaOfInterest) { [weak self] (parameters, error) in
+            if let parameters = parameters {
+                self?.offlineMapTask(offlineMapTask, didCreate: parameters)
+            } else if let error = error {
+                self?.offlineMapTask(offlineMapTask, didFailToCreateParametersWith: error)
+            }
+        }
+    }
+    
+    func offlineMapTask(_ offlineMapTask: AGSOfflineMapTask, didCreate parameters: AGSGenerateOfflineMapParameters) {
+        let filename = parameters.referenceBasemapFilename as NSString
+        let name = filename.deletingPathExtension
+        let `extension` = filename.pathExtension
+        if let url = Bundle.main.url(forResource: name, withExtension: `extension`) {
+            let message: String = {
+                let displayName = Bundle.main.object(forInfoDictionaryKey: "CFBundleDisplayName") as! String
+                return String(format: "“%@” has a local version of the basemap used by the map. Would you like to use the local or the online basemap?", displayName)
+            }()
+            let alertController = UIAlertController(title: "Choose Basemap", message: message, preferredStyle: .actionSheet)
+            let offlineAction = UIAlertAction(title: "Local", style: .default) { (_) in
+                parameters.referenceBasemapDirectory = url.deletingLastPathComponent()
+                self.takeMapOffline(task: offlineMapTask, parameters: parameters)
+            }
+            alertController.addAction(offlineAction)
+            let onlineAction = UIAlertAction(title: "Online", style: .default) { (_) in
+                self.takeMapOffline(task: offlineMapTask, parameters: parameters)
+            }
+            alertController.addAction(onlineAction)
+            let cancelAction = UIAlertAction(title: "Cancel", style: .cancel) { (_) in
+                self.generateButtonItem.isEnabled = true
+                self.mapView.isUserInteractionEnabled = true
+            }
+            alertController.addAction(cancelAction)
+            alertController.preferredAction = offlineAction
+            alertController.popoverPresentationController?.barButtonItem = generateButtonItem
+            present(alertController, animated: true)
+        } else {
+            takeMapOffline(task: offlineMapTask, parameters: parameters)
+        }
+    }
+    
+    func offlineMapTask(_ offlineMapTask: AGSOfflineMapTask, didFailToCreateParametersWith error: Error) {
```

Comment:
The `offlineMapTask` parameter isn't used here, is it? Seems odd to pass it in.

Actually, though I know this has been approved and this is just a cursory review, I think the pattern here obscures the ArcGIS pattern a bit. It's helpful to see Map, OfflineMapTask and GenerateOfflineMapJob together to understand that they relate. This pattern passes the Task around, doesn't even hold on to the Map (so there's some optional unwrapping that could be avoided by a map property), and has a property for the Job.

I can open an issue to clear it up if you like. Overall this sample buries the ArcGIS code deep within UI code.

Reply:
> The offlineMapTask parameter isn't used here, is it? Seems odd to pass it in.

It isn't, but I wanted to maintain symmetry between the `didCreate` and `didFailToCreate` methods. That works out nicely since they line up at the call site:

```swift
self?.offlineMapTask(offlineMapTask, didCreate: parameters)
self?.offlineMapTask(offlineMapTask, didFailToCreateParametersWith: error)
```

> This pattern passes the Task around, doesn't even hold on to the Map

The map is passed by passing the task (`AGSOfflineMapTask.onlineMap`).

> so there's some optional unwrapping that could be avoided by a map property

True, but there is only one place where the map is force unwrapped:

```swift
let offlineMapTask = AGSOfflineMapTask(onlineMap: mapView.map!)
```

 I'm not sure it's worth creating a map property for that one instance.

> I can open an issue to clear it up if you like.

Please do and link to this PR. I'll try to incorporate as many of your suggestions as I can.

Reply:
> I'm not sure it's worth creating a map property for that one instance.

Fair enough. I just really really don't like seeing `!` anywhere except on an `@IBOutlet`. It always worries me :)

> The map is passed by passing the task (AGSOfflineMapTask.onlineMap).

Also fair. I would like to see a property for the `AGSOfflineMapTask` though, but that's something to be done for the issue I'll create.

Reply:
> I would like to see a property for the `AGSOfflineMapTask`

We could certainly do that. One issue I see with having a property for the task is that it would be optional since the task is created on demand. By passing it as a parameter, I was able to keep all references to the task as non-optional.